modules/core/include/opencv2/core/utils/logger.hpp

```diff
@@ -59,7 +73,85 @@ CV_EXPORTS void writeLogMessage(LogLevel logLevel, const char* message);
 # endif
 #endif
 
+#define CV_LOGTAG_PTR_CAST(expr) static_cast<const cv::utils::logging::LogTag*>(expr)
 
+// CV_LOGTAG_EXPAND_NAME is intended to be re-defined (undef and then define again)
+// to allows logging users to use a shorter name argument when calling
+// CV_LOG_WITH_TAG or its related macros such as CV_LOG_INFO.
+//
+// This macro is intended to modify the tag argument as a string (token), via
+// preprocessor token pasting or metaprogramming techniques. A typical usage
+// is to apply a prefix, such as
+// ...... #define CV_LOGTAG_EXPAND_NAME(tag) cv_logtag_##tag
+//
+// It is permitted to re-define to a hard-coded expression, ignoring the tag.
+// This would work identically like the CV_LOGTAG_FALLBACK macro.
+//
+// Important: When the logging macro is called with tag being NULL, a user-defined
+// CV_LOGTAG_EXPAND_NAME may expand it into cv_logtag_0, cv_logtag_NULL, or
+// cv_logtag_nullptr. Use with care. Also be mindful of C++ symbol redefinitions.
+//
+// If there is significant amount of logging code with tag being NULL, it is
+// recommended to use (re-define) CV_LOGTAG_FALLBACK to inject locally a default
+// tag at the beginning of a compilation unit, to minimize lines of code changes.
+//
+#define CV_LOGTAG_EXPAND_NAME(tag) tag
+
+// CV_LOGTAG_FALLBACK is intended to be re-defined (undef and then define again)
+// by any other compilation units to provide a log tag when the logging statement
+// does not specify one. The macro needs to expand into a C++ expression that can
+// be static_cast into (cv::utils::logging::LogTag*). Null (nullptr) is permitted.
+#define CV_LOGTAG_FALLBACK nullptr
+
+// CV_LOGTAG_GLOBAL is the tag used when a log tag is not specified in the logging
+// statement nor the compilation unit. The macro needs to expand into a C++
+// expression that can be static_cast into (cv::utils::logging::LogTag*). Must be
+// non-null. Do not re-define.
+#define CV_LOGTAG_GLOBAL cv::utils::logging::internal::getGlobalLogTag()
+
+#define CV_LOG_WITH_TAG(tag, msgLevel, ...) \
+    for(;;) { \
+        const auto cv_temp_msglevel = (cv::utils::logging::LogLevel)(msgLevel); \
+        if (cv_temp_msglevel >= (CV_LOG_STRIP_LEVEL)) break; \
+        auto cv_temp_logtagptr = CV_LOGTAG_PTR_CAST(CV_LOGTAG_EXPAND_NAME(tag)); \
+        if (!cv_temp_logtagptr) cv_temp_logtagptr = CV_LOGTAG_PTR_CAST(CV_LOGTAG_FALLBACK); \
+        if (!cv_temp_logtagptr) cv_temp_logtagptr = CV_LOGTAG_PTR_CAST(CV_LOGTAG_GLOBAL); \
```

Comment:
I wrote this in the most verbose way in order to make its internal workings obvious, for code review (design review) purpose. 

The current code looks like it is doing null-pointer-coalescing (something available in C# as an operator), but there can be more logic added to it in between the lines. For example, it may perform additional log stripping checks. If it turns out the flexibility is not needed, we can discuss ways to simplify the code.

I am open to proposals that take away one or more "log tag injection mechanisms" - 

1. tag argument passed into the macro 
  - I believe it is worth keeping
2. "re-definable" `CV_LOGTAG_FALLBACK` macro 
  - good for getting existing OpenCV code to provide log tags while minimizing scope of code changes
  - bad because may confuse many C++ analysis and refactoring tools. Some tools do not support/allow re-defining macros.
3. ultimate `CV_LOGTAG_GLOBAL` macro
  - alternative: we can simply combine this functionality with `CV_LOGTAG_FALLBACK`, in other words, the default value for `CV_LOGTAG_FALLBACK` would be the global log tag.
    - but if we want to accept this alternative (getting rid of `CV_LOGTAG_GLOBAL`), first we need accept that macros can be re-defined.